samples/Rich-Cards-Bot/main.py

```diff
@@ -7,245 +7,46 @@
-        '5': [create_receipt_card], 'receipt card': [create_receipt_card],
-        '6': [create_signin_card], 'signin card': [create_signin_card],
-        '7': [create_thumbnail_card], 'thumbnail card': [create_thumbnail_card],
-        '8': [create_video_card], 'video card': [create_video_card],
-        '9': [create_adaptive_card, create_animation_card, create_audio_card, create_hero_card,
-              create_receipt_card, create_signin_card, create_thumbnail_card, create_video_card],
-        'all cards': [create_adaptive_card, create_animation_card, create_audio_card, create_hero_card,
-                      create_receipt_card, create_signin_card, create_thumbnail_card, create_video_card]
-    }
-
-    # Get the functions that will generate the card(s) for our response
-    # If the stripped response from the user is not found in our choice_dict, default to None
-    choice = choice_dict.get(response, None)
-    # If the user's choice was not found, respond saying the bot didn't understand the user's response.
-    if not choice:
-        not_found = await create_reply_activity(context.activity, 'Sorry, I didn\'t understand that. :(')
-        await context.send_activity(not_found)
-        return web.Response(status=202)
-    else:
-        for func in choice:
-            card = func()
-            response = await create_reply_activity(context.activity, '', card)
-            await context.send_activity(response)
-        return web.Response(status=200)
-
-
-async def handle_conversation_update(context: TurnContext) -> web.Response:
-    if context.activity.members_added[0].id != context.activity.recipient.id:
-        response = await create_reply_activity(context.activity, 'Welcome to the Rich Cards Bot!')
-        await context.send_activity(response)
-    return web.Response(status=200)
-
-
-async def unhandled_activity() -> web.Response:
-    return web.Response(status=404)
-
+user_state = UserState(memory)
+conversation_state = ConversationState(memory)
 
-async def request_handler(context: TurnContext) -> web.Response:
-    if context.activity.type == 'message':
-        return await handle_message(context)
-    elif context.activity.type == 'conversationUpdate':
-        return await handle_conversation_update(context)
-    else:
-        return await unhandled_activity()
 
+dialog = MainDialog()
+bot = RichCardsBot(conversation_state, user_state, dialog)
 
-async def messages(req: web.web_request) -> web.Response:
+async def messages(req: web.Request) -> web.Response:
     body = await req.json()
     activity = Activity().deserialize(body)
     auth_header = req.headers['Authorization'] if 'Authorization' in req.headers else ''
+    async def aux_func(turn_context):
+        await bot.on_turn(turn_context)
+
     try:
-        return await ADAPTER.process_activity(activity, auth_header, request_handler)
+        await ADAPTER.process_activity(activity, auth_header, aux_func)
```

Comment:
I didn't see an `adapter.onTurnError()` equivalent in the sample, is this in the library's implementation?

Reply:
Example in 4.4 Core Bot JS: https://github.com/microsoft/BotBuilder-Samples/blob/447028bd6fb18cdc81ef19d5e830027ce4826d5e/samples/javascript_nodejs/13.core-bot/index.js#L28-L38

Reply:
There is no current functionality for this, I opened this issue on the matter: #167